### PR TITLE
Add analysis on SonarCloud.

### PR DESCRIPTION
**Pull request by bellingard**
@Mefl As discussed on [SonarSource Community forum](https://community.sonarsource.com/t/adding-a-source-code-analysis-for-iot/22143/7), here's the PR to activate SonarCloud analysis on the repo, using Travis-CI.

The following steps should be enough to make this work:
1. Activate Travis-CI (https://travis-ci.com) for your repo
   * Follow steps 1, 2 and 3 of [the tutorial](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci-using-github)
2. On the [Settings page of your repo in Travis](https://travis-ci.com/github/covid-response-projects/covid-respirator/settings), set a SonarCloud token as a SONAR_TOKEN environment variable
   * You can generate a SonarCloud token on https://sonarcloud.io/account
3. Merge this PR into your master branch, this will kick in a build on Travis and a couple of minutes later you will have the analysis available on [project homepage](https://sonarcloud.io/dashboard?id=covid-response-projects_covid-respirator)

**Important note**: for the SonarCloud analysis to run, it's mandatory that the Platform IO build succeeds because it needs information from the compiler. Currently, your master branch is broken (see [the latest run](https://github.com/covid-response-projects/covid-respirator/runs/540293426)), so you should fix this prior to merging this PR - otherwise you won't get results.

Please feel free to ping me if anything does not work as expected!

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+language: python
+python:
+    - "2.7"
```
Comment (dsferruzza) [inline .travis.yml]: Isn't Python 2.x EOL?

Comment (bellingard) [reply .travis.yml]: Yes, Python 2.x is EOL, but I don't think this really matters for now. This is [the example that PlatformIO gives for Travis-CI integration](https://docs.platformio.org/en/latest/integration/ci/travis.html) - so we simply followed that.

**Comment by bellingard** [comment]
@abbas-sabra-sonarsource Looks like they plan to drop PlatformIO and use [arduino-cli](https://github.com/arduino/arduino-cli) instead. Would our build wrapper work in this case?
If we support this compiler, looks like we could download the CLI and use it just like this in Travis CI: https://github.com/arduino-libraries/ArduinoIoTCloud/pull/66/files

**Comment by abbas-sabra-sonarsource** [comment]
@bellingard It shouldn't be an issue for us since the same compiler (avr-g++) will be used under the hood. The platform that is calling the compiler doesn't matter for our analyzer as long as the compiler is supported.

When they move to arduino-cli these two lines should be changed from:
```
        platformio run -t clean
        build-wrapper-linux-x86-64 --out-dir bo platformio run --verbose
```
to
```
        **arduino-cli clean command**
        build-wrapper-linux-x86-64 --out-dir bo  **arduino-cli build command** 
```

**Comment by bellingard** [comment]
Great, thanks @abbas-sabra-sonarsource for the confirmation that we do support this compiler. On top of the changes you mentioned, I will have to update the rest of the Travis file to download and install arduino-cli.

@dsferruzza: so up to you to decide what to do with this! Whatever you choose, we can support you. Just tell me once you know!

**Comment by dsferruzza** [comment]
Hi!
Sorry for the delay, we had a lot to handle during the past days...

From my perspective, there are several issues we have to solve to make it work:

1. We now use [arduino-cli](https://github.com/arduino/arduino-cli) instead of PlatformIO because we are now targeting STM32 boards (not Arduino Uno anymore). I believe this would change the way we need to build the code for Sonar. You can find a working build process in `.github/workflows/arduino-cli.yml`.
2. For now, we have 2 programs in the codebase: the real program, and a qualification firmware that can be used to check that sensors and actuators act normally. Because arduino-cli really wants to build everything (and ends up crashing because there are 2 `loop()` functions), we had to use a conditional macro to be able to choose which program to build; the switch is in `config.h` (this is dirty but it delivers for now). I believe we should check both programs with Sonar; how do you think we shall proceed? (Again, you can have a look to `.github/workflows/arduino-cli.yml` where we build both of them)

How do you think we could solve these?

**Comment by dsferruzza** [comment]
Ping @bellingard

**Comment by abbas-sabra-sonarsource** [comment]
Hello @dsferruzza,

Sorry, it took me a long time to get back to you, I wasn't notified when you moved to arduino-cli.
I had a look at .github/workflows/arduino-cli.yml. If you have multiple build configurations to build your project, you can have separate analysis for every configuration so you make sure you cover them all. I created in this PR two `build.sh` one for "Build production HW1" and one for "Build production HW2". You can see the two projects analyzed [here](https://sonarcloud.io/organizations/abbas-sabra-sonarsource/projects). 
of course, you can extend this methodology to the other configurations if needed. All you have to do is to pass the new build configuration to build-wrapper and configure the new project with the sonar-project.properties.

I have faced one problem that you need to be aware of if you decided to use SonarCloud to analyze C and C++. Arduino-cli copies source files to temporary locations before building them: [Link](https://github.com/arduino/arduino-cli/issues/143). This is problematic for us since we analyze the files that were built in the project directory. As a workaround to solve this problem I have added --`build-path $PWD/*buildOutputFile` to all the `arduino-cli compile` commands passed to build-wrapper. This makes sure the source files aren’t copied outside the project directory, but still, the copied files are the one who is built/analyzed so directory of the analyzed files will look different on SonarCloud. Also, there is an issue if you give --build-path a relative directory; you have to give the absolute path or the build will fail. Issue reported [here](https://github.com/arduino/arduino-builder/issues/235#issuecomment-440557544).

- Here are the results of [Build production HW1](https://sonarcloud.io/dashboard?id=covid_HW1).
- Here are the results of [Build production HW2](https://sonarcloud.io/dashboard?id=covid_HW2).
- The analyzer found a bug. You can review it [here](https://sonarcloud.io/project/issues?id=covid_HW2&open=AXGZDPfwXrU23ba2qP3-&resolved=false&types=BUG).
- Also if you are interested in Following the Misra standard we provide some rules to enforce it that aren’t enabled by default.
- Fabrice comments about activating Travis and SONAR_TOKEN still need to be done from your side.

Don’t hesitate to contact me if anything is unclear or if you face a configuration issue later on.

**Comment by jabby** [comment]
:pray:  @bellingard @abbas-sabra-sonarsource  Thanks for your work, I merged another PR based on your work. It works the same as your work.  
See our report here https://sonarcloud.io/organizations/makers-for-life/projects/

I just notice there is code duplication due to how we build all projects with arduino-cli. I will work on this point during next few days. If you have any idea, feel free to ping me here.

Thanks again.

**Comment by bellingard** [comment]
Hey @jabby!
if you set `sonar.sources=src` instead of `sonar.sources=.` in your Travis YML script, this will exclude the `build/` folder (which contains all the duplications).

**Comment by abbas-sabra-sonarsource** [comment]
Hello @bellingard @jabby,

If they set `sonar.sources=src` the analysis will stop working. sonar.sources should point to a directory of the analyzed files which is `build`. Setting `sonar.sources=build` should solve the issue.

**Comment by bellingard** [comment]
Indeed @abbas-sabra-sonarsource, you're right. Given that arduino-cli copies the files in the `build` folder for its own needs, the build-wrapper will get insights on those copies and not the original source files. :-(
This is unfortunate because in SonarCloud, the dev team won't see the right file paths, nor the blame information and most importantly issues won't be auto-assigned to the relevant persons. It's an unexpected situation, but at least the analysis runs and gives feedback on the source code!

**Pull request closed without merging**